osu.Game/Screens/Play/HUD/KeyCounterDisplay.cs

```diff
@@ -0,0 +1,113 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using osu.Framework.Allocation;
+using osu.Framework.Bindables;
+using osu.Framework.Extensions.IEnumerableExtensions;
+using osu.Framework.Graphics;
+using osu.Framework.Graphics.Containers;
+using osu.Framework.Input.Events;
+using osu.Game.Configuration;
+using osuTK;
+
+namespace osu.Game.Screens.Play.HUD
+{
+    /// <summary>
+    /// A flowing display of all gameplay keys. Individual keys can be added using <see cref="InputTrigger"/> implementations.
+    /// </summary>
+    public abstract partial class KeyCounterDisplay : Container<KeyCounter>
+    {
+        /// <summary>
+        /// Whether the key counter should be visible regardless of the configuration value.
+        /// This is true by default, but can be changed.
+        /// </summary>
+        public Bindable<bool> AlwaysVisible { get; } = new Bindable<bool>(true);
+
+        public Bindable<bool> IsCounting { get; } = new BindableBool(true);
+
+        protected readonly Bindable<bool> ConfigVisibility = new Bindable<bool>();
+
+        protected abstract void UpdateVisibility();
+
+        private Receptor? receptor;
+
+        private readonly Type[] acceptedTypes;
+
+        protected KeyCounterDisplay(params Type[] acceptedTypes)
+        {
+            this.acceptedTypes = acceptedTypes;
+        }
+
+        public void SetReceptor(Receptor receptor)
+        {
+            if (this.receptor != null)
+                throw new InvalidOperationException("Cannot set a new receptor when one is already active");
+
+            this.receptor = receptor;
+        }
+
+        public abstract void AddTrigger(InputTrigger trigger);
+
+        public void AddTriggerRange(IEnumerable<InputTrigger> triggers) => triggers.ForEach(AddTrigger);
+
+        public override void Add(KeyCounter counter)
+        {
+            if (!checkType(counter))
+                throw new InvalidOperationException($"{counter.GetType()} is not a supported counter type. (hint: you may want to use {nameof(AddTrigger)} instead.)");
+
+            base.Add(counter);
+            counter.IsCounting.BindTo(IsCounting);
+        }
```

Comment:
I started to apply some final cleanups to this pull but this is where I get stuck. Why is this here still? As in, why is any subclass still allowed to arbitrarily add any `KeyCounter` if the type isn't matching?

Like this is so weird for me, because `KeyCounterDisplay` is apparently a `Container`, but then `DefaultKeyCounterDisplay` is setting `InternalChild` anyway, and thus is just nullifying the fact that its base class is a container.

I would like to see the process of creating the individual key counters and placing them in the key counter display's layout entirely encapsulated *within* the key counter display, because it shouldn't ever *make sense* to mix and match key counter implementations. It should be fundamentally impossible to misuse the component. This would be done by making `KeyCounter` a plain `CompositeDrawable`, with the expectation that implementors of `AddTrigger()` should in response create the `KeyCounter` they want and place it in the layout of the key counter display according to expectation.

Reply:
I did consider guarding against this further but wasn't sure it was worth the effort. Don't disagree with direction though.

Reply:
That said, I did initially try to refactor to a `CompositeDrawable` but it gets messy since `CompositeDrawable` doesn't support `Content` overrides.

As can be seen by https://github.com/ppy/osu/pull/22654/commits/44297a7d0a01e44f24d2860cdd4ce2584988e341.

So I'm not 100%, see what you think I guess.